File: python/100_Same_Tree.py
#!/usr/bin/python
import unittest



# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def isSameTree(self, p, q):
    	if p == q == None: # not enough, since different addresses may have the same value
    		return True
    	elif (p == None and q != None) or (p != None and q == None):
    		return False
    	elif p.val == q.val:
    		resLeft = self.isSameTree(p.left, q.left)
    		resRight = self.isSameTree(p.right, q.right)
    		if resRight and resLeft:
    			return True
    	return False




class TestSolution(unittest.TestCase):
    def setUp(self):
        self.s = Solution()

    def test_method(self):
        """Such as self.assertEqual, self.assertTrue"""
        #self.assertEqual(self.s.)
        # [10,5,15], [10,5,null,null,15]
        # 0
        pass

def main():
	s = Solution()

	a1 = TreeNode(10)
	a2 = TreeNode(5)
	a3 = TreeNode(15)
	a1.left = a2
	a1.right = a3

	b1 = TreeNode(10)
	b2 = TreeNode(5)
	b3 = TreeNode(None)
	b4 = TreeNode(None)
	b5 = TreeNode(15)
	b1.left = b2
	b1.right = b3
	b2.left = b4
	b2.right = b5

	print s.isSameTree(a1, b1)


if __name__ == "__main__":
    #unittest.main()
    main()

File: c++/100_Same_Tree.cpp
#include <iostream>
#include <vector>
#include <map>
#include <string>
#include <algorithm>

using namespace std;



struct TreeNode {
	int val;
	TreeNode *left;
	TreeNode *right;
	TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
 
class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
    	if (!p and !q) return true;
    	else if ((!p && q) || (p && !q)) return false;
    	else if (p->val == q->val) {
    		bool isLeft = isSameTree(p->left, q->left);
    		bool isRight = isSameTree(p->right, q->right);
    		if (isLeft and isRight)
    			return true;
    	}
    	return false;
    }
};

// class Solution {
// public:
//     bool isSameTree(TreeNode *p, TreeNode *q) {
//         if(!p && !q) return true;
//         if((!p && q) || (p && !q)) return false;
//         if(p->val != q->val) return false;
//         return isSameTree(p->left,q->left) && isSameTree(p->right,q->right);
//     }
// };


int main() {
	Solution s;

	TreeNode* a1 = new TreeNode(1);
	TreeNode* a3 = new TreeNode(1);
	a1->left = NULL;
	a1->right = a3;

	TreeNode* b1 = new TreeNode(1);
	TreeNode* b3 = new TreeNode(1);
	b1->left = NULL;
	b1->right = b3;
	cout << s.isSameTree(a1, b1) << endl;
}